proposals/0003-health-checks.md

```diff
@@ -0,0 +1,103 @@
+# Service Healthchecks
+
+* Proposal: [MP-0003](0003-health-checks.md)
+* Authors: [Clement Escoffier](https://github.com/cescoffier), [Heiko Rupp](https://github.com/pilhuhn), [Heiko Braun] (https://github.com/heiko-braun)
+
+* Status: **Awaiting review**
+
+*During the review process, add the following fields as needed:*
+
+* Decision Notes: [Discussion thread topic covering the  Rationale](https://groups.google.com/forum/#!forum/microprofile), [Discussion thread topic with additional Commentary](https://groups.google.com/forum/#!forum/microprofile)
+
+## Introduction
+
+This document describes a protocol (wireformat, semantics and possible forms of interactions) between system components that need to determine the “liveliness” of computing nodes in a bigger architecture.
+
+
+Mailinglist thread: [Discussion thread topic for that proposal](https://groups.google.com/forum/#!topic/microprofile/jIZAKiu76ys)
+
+## Motivation
+
+Health checks are used to probe the state of a computing node from another other machine (i.e. kubernetes service controller) with the primary target being cloud infrastructure environments where automated processes maintain the state of computing nodes. In this scenario health checks are used to determine if a computing node needs to be discarded (terminated, shutdown) and eventually replaced by another (healthy) instance.
```

Comment:
Like the proposal 👍 

I would like to enhance it or give it another dimention, lets call it '**Smart Health-Check**' or '**Health-Check 360**'. What I mean by this is, consider following scenario:

A given service/microservice could be dependent on multiple downstream/remote applications (or we call them another point of failure for the microservice). In most use cases a service connects to a DB or another service(s), such system design leads to failing the service in case of any of the given dependent service is failed (their individual health-check says 'DOWN'). Suppose a service called MyService depends on PostgresDB and downstream services called AServ, BServ:
```bash
MyServ -> PostgresDB
              -> AServ
              -> BServ
```
We could define the response of Health-Check for MyServ as:

 1.  UP : if PostgresDB, AServ & BServ are all UP.
 2. UP : if PostgresDB is UP but  AServ & BServ are all DOWN. 
 3. DOWN : otherwise

above is not the final list of strategies however these could depend on the use cases. So such strategies should be configurable at the time of defining/designing the service.

*If 360 view of the microservice is not taken into consideration then we are telling half truth.*

What I meant to say is that Health-Check doesn't mean a service's own health however it depends on the environment/topology it runs on.